### PR TITLE
Feature 2709 mvmode multithreshradii

### LINKED ISSUE #2709: Enhance Multivariate MODE to support multiple convolution radii and thresholds
## Describe the Enhancement ##

- Allow multiple convolution radii and thresholds without quilting.
- Currently disallowed in multivariate mode.
- Can also do this with multiple METplus configurations (I think) in case this does not get done.
- Possibly also allow quilting if it is trivial to do so.

### Time Estimate ###
40 hours

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
- [x] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
Work must be completed by 12/30 for NRL METplus funding

### Funding Source ###
NRL METplus 7730022

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Review default **alert** labels
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Milestone and Projects ###
- [ ] Select **Milestone** as the next official version or **Backlog of Development Ideas**
- [ ] For the next official version, select the **MET-X.Y.Z Development** project

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose)

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [x] Complete the development and test your changes.
- [x] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [x] Add/update documentation.
- [x] Push local changes to GitHub.
- [x] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [x] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the next official version
Select: **MET-X.Y.Z Development** project for development toward the next official release
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by davidalbo**
## Expected Differences ##
quilting and non-quilting options are now supported. 

Without quilting all inputs (forecast and obs) need to have the same number N of convolution radii and thresholds, with one mvmode run for the first set, one for the second set, etc. for a total of N runs.

With quilting all inputs must have the same number of convolution radii N, and all inputs must have the same number of convolution thresholds M, but N and M can differ. N*M mvmode runs are performed.

- [ ] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>

- [ ] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>

## Pull Request Testing ##

- [ ] Describe testing already performed for these changes:</br>
I ran several tests on seneca on the same case I've used for other MvMode changes.  Tested with and without quilting.  The tests can be found here:  /d1/personal/dave/mvmode_november_test.  The scripts that I ran are:  run-multi.bsh  run-quilt.bsh check-multi.bsh.
Comments in these files show what I was testing.  I did kind of make up some of the convolution settings (not scientific).

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
I'd recommend starting with these tests to see if you like the logging and output file names and content.  Everything is local to /d1/personal/dave/mvmode_november_test.

- [ ] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**  I hope no errors or warning happen, my changes were pretty minor to the user document.

- [ ] Do these changes include sufficient testing updates? **[Maybe]**  @hertneky you are wecome to create a better more meaningful scientific test or two.

- [ ] Will this PR result in changes to the MET test suite? **[Maybe]**</br>
I'm wondering if the version which gets written to the output files will cause the test suite to fail, because the version has changed to v12.1.  We'll find out.
Also, should we add a test that has mvmode with multiple conv. thresh/radii to the unit tests?

- [ ] Will this PR result in changes to existing METplus Use Cases? **[Maybe?]**</br>
If **yes**, create a new **Update Truth** [METplus issue](https://github.com/dtcenter/METplus/issues/new/choose) to describe them.
Will this version change to  v12.1  which is written to output files cause problems with existing METplus Use Cases?

- [ ] Do these changes introduce new SonarQube findings? **[?]**</br>  How do I test for that?

- [ ] Please complete this pull request review by **[Dec 31]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **MET-X.Y.Z Development** project for official releases
- [ ] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by JohnHalleyGotway** [comment]
On Dec 11, 2024, I compiled the feature branch for this PR on seneca for @hertneky and @JohnHalleyGotway to test.

@hertneky please test the executables in this directory:
`seneca:/d1/projects/MET/MET_pull_requests/met-12.1.0/beta1/MET-feature_2709_mvmode_multithreshradii/bin`

I did this using the following commands:
```
# On seneca
cd /d1/projects/MET/MET_pull_requests/met-12.1.0/beta1
runas met_test # Compile all PR branches as a shared user
git clone https://github.com/dtcenter/MET MET-feature_2709_mvmode_multithreshradii
cd MET-feature_2709_mvmode_multithreshradii
git checkout feature_2709_mvmode_multithreshradii
source internal/scripts/environment/development.seneca
./configure --prefix=`pwd` --enable-all
make install test
```

**Comment by hertneky** [comment]
@JohnHalleyGotway Got it, thanks. I will do some testing tomorrow.

File: docs/Users_Guide/mode.rst [hunk 1]
```diff
@@ -319,7 +323,7 @@ The **conv_radius** entry defines the radius of the circular convolution applied
 
 The **conv_thresh** entry specifies the threshold values to be applied to the convolved field to define objects. By default, objects are defined using a convolution threshold of 5.0. Multiple convolution thresholds may be specified as an array (e.g. **conv_thresh = [ >=5.0, >=10.0, >=15.0 ];)**.
 
-Multiple convolution radii and thresholds and processed using the logic defined by the **quilt** entry.
+Multiple convolution radii and thresholds and processed using the logic defined by the **quilt** entry.  The logic specific to multivariate mode is described in the multivariate mode section above.
```
Comment (hertneky) [inline docs/Users_Guide/mode.rst]: Multiple convolution radii and thresholds **are** processed.

**Review by hertneky — DISMISSED** [review]
The documentation looks good to me - just one small typo.

Testing done:

Quilting=False
equal inputs for conv radii/thresh - passed, produced expected results
unequal inputs for conv radii/thresh - passed, produced informative error when unequal inputs present, which isn't allowed.

Quilting=True
Equal inputs for radii and equal inputs for thresh - Passed, produced expected results
Unequal number of radii - Passed, produced informative error
Unequal number of thresh - Passed, produced informative error
Unequal number of thresh and merge_thresh - Passed, produced informative error

Let me know of any other tests that might be useful, but it looks good so far!

File: docs/Users_Guide/mode.rst (hunk 1 above)
Comment (JohnHalleyGotway) [reply docs/Users_Guide/mode.rst]: @hertneky, I fixed this typo.

**Comment by JohnHalleyGotway** [comment]
Note that the SonarQube code smells are reduced from 18,277 in develop to 18,072 in this feature branch.

**Review by JohnHalleyGotway — APPROVED** [review]
@davidalbo I approve of these changes. I did some work over the last few days to drive down the number of SonarQube code smells, and I'm happy with the result. Ideally we'd also add a test to demonstrate running mv mode using multiple radii or thresholds.

So you could either merge now... or if it's easy to extend an existing unit test to supply more than one radius and/or thresholds, please do so.

**Comment by davidalbo** [comment]
> So you could either merge now... or if it's easy to extend an existing unit test to supply more than one radius and/or thresholds, please do so.

Due to funding halt, I'll merge now, and we can consider another test or extension of existing test later.

**Pull request merged**